src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,141 @@
+#pragma warning disable IDE0051
+
+using Neo.Cryptography;
+using Neo.Cryptography.ECC;
+using Neo.Ledger;
+using Neo.Network.P2P.Payloads;
+using Neo.Persistence;
+using Neo.SmartContract.Manifest;
+using Neo.VM.Types;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Neo.SmartContract.Native.Oracle
+{
+    public sealed partial class OracleContract : NativeContract
+    {
+        private const int MaxUrlLength = 256;
+        private const int MaxFilterLength = 128;
+        private const int MaxCallbackLength = 32;
+        private const int MaxUserDataLength = 512;
+
+        private const byte Prefix_NodeList = 8;
+        private const byte Prefix_RequestId = 9;
+        private const byte Prefix_Request = 7;
+        private const byte Prefix_IdList = 6;
+
+        private const long OracleRequestPrice = 0_50000000;
+
+        public override int Id => -4;
+        public override string Name => "Oracle";
+
+        internal OracleContract()
+        {
+            Manifest.Features = ContractFeatures.HasStorage | ContractFeatures.Payable;
+        }
+
+        [ContractMethod(0, CallFlags.AllowModifyStates)]
+        private void Finish(ApplicationEngine engine)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            OracleResponse response = tx.Attributes.OfType<OracleResponse>().First();
+            StorageKey key = CreateStorageKey(Prefix_Request).Add(response.Id);
+            OracleRequest request = engine.Snapshot.Storages[key].GetInteroperable<OracleRequest>();
+            StackItem userData = BinarySerializer.Deserialize(request.UserData, engine.MaxStackSize, engine.MaxItemSize, engine.ReferenceCounter);
+            engine.CallFromNativeContract(null, request.CallbackContract, request.CallbackMethod, request.Url, userData, response.Success, response.Result);
+        }
+
+        [ContractMethod(0_01000000, CallFlags.AllowStates)]
+        public ECPoint[] GetOracleNodes(StoreView snapshot)
+        {
+            return snapshot.Storages[CreateStorageKey(Prefix_NodeList)].GetInteroperable<NodeList>().ToArray();
+        }
+
+        public OracleRequest GetRequest(StoreView snapshot, ulong id)
+        {
+            return snapshot.Storages.TryGet(CreateStorageKey(Prefix_Request).Add(id))?.GetInteroperable<OracleRequest>();
+        }
+
+        public IEnumerable<OracleRequest> GetRequests(StoreView snapshot)
```

Comment:
This method is not used

Reply:
This method is used by oracle nodes. (which has not been implemented)